File: EdT_ESP2/src/services/matiereService.ts
import api from "./api";
import type { CreateMatiereDto, MatiereDto } from "../types";

const MATIERES_BASE = "/matieres";

export const matiereService = {
  // Créer une matière
  create: async (data: CreateMatiereDto): Promise<MatiereDto> => {
    const response = await api.post<MatiereDto>(MATIERES_BASE, data);
    return response.data;
  },

  // Liste des matières
  getAll: async (): Promise<MatiereDto[]> => {
    const response = await api.get<MatiereDto[]>(MATIERES_BASE);
    return response.data;
  },

  // ✅ Liste des matières par département
  getByDepartement: async (departementId: number): Promise<MatiereDto[]> => {
    const response = await api.get<MatiereDto[]>(MATIERES_BASE, {
      params: { departementId },
    });
    return response.data;
  },

  // Détails d'une matière
  getByCode: async (code: string): Promise<MatiereDto> => {
    const response = await api.get<MatiereDto>(`${MATIERES_BASE}/${code}`);
    return response.data;
  },

  // Modifier une matière
  update: async (code: string, data: CreateMatiereDto): Promise<MatiereDto> => {
    const response = await api.put<MatiereDto>(`${MATIERES_BASE}/${code}`, data);
    return response.data;
  },

  // Supprimer une matière
  delete: async (code: string): Promise<void> => {
    await api.delete(`${MATIERES_BASE}/${code}`);
  },
};

export default matiereService;

File: edt_mobile/lib/services/matiere_service.dart
import 'api_client.dart';
import '../models/matiere.dart';

class MatiereService {
  final _dio = ApiClient().dio;

  /// Get all matières → GET /matieres
  Future<List<MatiereDto>> getAll() async {
    final response = await _dio.get('/matieres');
    return (response.data as List)
        .map((e) => MatiereDto.fromJson(e))
        .toList();
  }

  /// Get matières by department → GET /matieres?departementId=X
  Future<List<MatiereDto>> getByDepartement(int departementId) async {
    final response = await _dio.get('/matieres', queryParameters: {
      'departementId': departementId,
    });
    return (response.data as List)
        .map((e) => MatiereDto.fromJson(e))
        .toList();
  }
}
